### 데이터셋 불러오기

In [ ]:
from sklearn.datasets import load_breast_cancer

cancer = load_breast_cancer()

print("유방암 데이터셋: \n{}".format(cancer.keys()))
print("유방암 데이터의 형태: {}".format(cancer.data.shape))

유방암 데이터셋: 
dict_keys(['data', 'target', 'target_names', 'DESCR', 'feature_names', 'filename'])
유방암 데이터의 형태: (569, 30)


유방암 데이터는 30개의 특성을 가지고 있으며 569개의 데이터를 가지고 있습니다.

In [ ]:
print("특성 이름:\n{}".format(cancer.feature_names))

특성 이름:
['mean radius' 'mean texture' 'mean perimeter' 'mean area'
 'mean smoothness' 'mean compactness' 'mean concavity'
 'mean concave points' 'mean symmetry' 'mean fractal dimension'
 'radius error' 'texture error' 'perimeter error' 'area error'
 'smoothness error' 'compactness error' 'concavity error'
 'concave points error' 'symmetry error' 'fractal dimension error'
 'worst radius' 'worst texture' 'worst perimeter' 'worst area'
 'worst smoothness' 'worst compactness' 'worst concavity'
 'worst concave points' 'worst symmetry' 'worst fractal dimension']


In [ ]:
print("데이터 정보:\n{}".format(cancer.DESCR))

데이터 정보:
.. _breast_cancer_dataset:

Breast cancer wisconsin (diagnostic) dataset
--------------------------------------------

**Data Set Characteristics:**

    :Number of Instances: 569

    :Number of Attributes: 30 numeric, predictive attributes and the class

    :Attribute Information:
        - radius (mean of distances from center to points on the perimeter)
        - texture (standard deviation of gray-scale values)
        - perimeter
        - area
        - smoothness (local variation in radius lengths)
        - compactness (perimeter^2 / area - 1.0)
        - concavity (severity of concave portions of the contour)
        - concave points (number of concave portions of the contour)
        - symmetry 
        - fractal dimension ("coastline approximation" - 1)

        The mean, standard error, and "worst" or largest (mean of the three
        largest values) of these features were computed for each image,
        resulting in 30 features.  For instance, field 3 is Mean R

### 데이터 전처리

In [ ]:
import numpy as np
from sklearn.preprocessing import MinMaxScaler

# 실행할 때마다 같은 결과를 출력하기 위해 설정하는 부분입니다.
np.random.seed(0)

x = cancer.data
y = cancer.target

# x에 대해 최대최소 정규화 실시

scaler = MinMaxScaler()
scaler.fit(x)

x_scaled = scaler.transform(x)

In [ ]:
from sklearn.preprocessing import StandardScaler

# 실행할 때마다 같은 결과를 출력하기 위해 설정하는 부분입니다.
np.random.seed(0)

x = cancer.data
y = cancer.target

# x에 대해 최대최소 정규화 실시

scaler1 = StandardScaler()
scaler1.fit(x)

x_scaled1 = scaler1.transform(x)

In [ ]:
# 정규화 후, 데이터 셋 속성 출력

print('정규화 전 특성별 최소값 : \n', x.min(axis=0))
print('정규화 전 특성별 최소값 : \n', x.max(axis=0))

print('정규화 후 특성별 최소값 : \n', x_scaled.min(axis=0))
print('정규화 후 특성별 최소값 : \n', x_scaled.max(axis=0))

정규화 전 특성별 최소값 : 
 [6.981e+00 9.710e+00 4.379e+01 1.435e+02 5.263e-02 1.938e-02 0.000e+00
 0.000e+00 1.060e-01 4.996e-02 1.115e-01 3.602e-01 7.570e-01 6.802e+00
 1.713e-03 2.252e-03 0.000e+00 0.000e+00 7.882e-03 8.948e-04 7.930e+00
 1.202e+01 5.041e+01 1.852e+02 7.117e-02 2.729e-02 0.000e+00 0.000e+00
 1.565e-01 5.504e-02]
정규화 전 특성별 최소값 : 
 [2.811e+01 3.928e+01 1.885e+02 2.501e+03 1.634e-01 3.454e-01 4.268e-01
 2.012e-01 3.040e-01 9.744e-02 2.873e+00 4.885e+00 2.198e+01 5.422e+02
 3.113e-02 1.354e-01 3.960e-01 5.279e-02 7.895e-02 2.984e-02 3.604e+01
 4.954e+01 2.512e+02 4.254e+03 2.226e-01 1.058e+00 1.252e+00 2.910e-01
 6.638e-01 2.075e-01]
정규화 후 특성별 최소값 : 
 [0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
 0. 0. 0. 0. 0. 0.]
정규화 후 특성별 최소값 : 
 [1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1. 1.
 1. 1. 1. 1. 1. 1.]


In [ ]:
from sklearn.model_selection import train_test_split

x_train, x_test, y_train, y_test = train_test_split(x_scaled, y, test_size=0.3, random_state=0)

### 딥러닝 모델 제작

In [ ]:
from tensorflow.keras.layers import Dense,SimpleRNN,LSTM,Dropout
from tensorflow.keras import models
from tensorflow.keras.optimizers import RMSprop
import matplotlib.pyplot as plt

model = models.Sequential()
model.add(Dense(32, input_dim = 30, activation='relu'))
model.add(Dense(32, activation='relu'))
model.add(Dense(1, activation='sigmoid'))



model.compile(optimizer=RMSprop(),loss='binary_crossentropy',metrics=['accuracy'])

In [ ]:
history = model.fit(x_train,y_train,epochs=30,batch_size=64,validation_data=(x_test, y_test))

Train on 398 samples, validate on 171 samples
Epoch 1/30
398/398 [==============================] - 0s 68us/sample - loss: 0.0573 - accuracy: 0.9874 - val_loss: 0.0881 - val_accuracy: 0.9766
Epoch 2/30
398/398 [==============================] - 0s 48us/sample - loss: 0.0574 - accuracy: 0.9824 - val_loss: 0.0851 - val_accuracy: 0.9766
Epoch 3/30
398/398 [==============================] - 0s 55us/sample - loss: 0.0525 - accuracy: 0.9849 - val_loss: 0.1033 - val_accuracy: 0.9357
Epoch 4/30
398/398 [==============================] - 0s 68us/sample - loss: 0.0548 - accuracy: 0.9824 - val_loss: 0.0878 - val_accuracy: 0.9766
Epoch 5/30
398/398 [==============================] - 0s 58us/sample - loss: 0.0546 - accuracy: 0.9824 - val_loss: 0.0888 - val_accuracy: 0.9825
Epoch 6/30
398/398 [==============================] - 0s 62us/sample - loss: 0.0529 - accuracy: 0.9874 - val_loss: 0.1067 - val_accuracy: 0.9357
Epoch 7/30
398/398 [==============================] - 0s 59us/sample - loss: 0.0560 

In [ ]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
import numpy
import tensorflow as tf
  
# seed 값 생성
np.random.seed(3)
tf.random.set_seed(3)
  
  
# 모델의 설정
model = Sequential()
model.add(Dense(12, input_dim=8, activation='relu'))
model.add(Dense(8, activation='relu'))
model.add(Dense(1, activation='sigmoid'))
  
# 모델 컴파일
model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])
  
# 모델 실행
model.fit(X, y, epochs=200, batch_size=10)
  
# 결과 출력
print("\n Accuracy: %.4f" % (model.evaluate(X, y)[1]))